plutus-core/plutus-ir/src/PlutusIR/Transform/NewLetFloat.hs

```diff
@@ -0,0 +1,493 @@
+-- This subset contains only the marks of the floatable lets.
+type Marks = Scope
+
+-- | A "naked" let, without its inTerm.
+-- We use this structure
+-- 1) to determine if a let is floatable/unfloatable. See 'floatable'.
+-- 2) to calculate freevars/tyvars of a let.
+-- 2) to store it in the 'FloatTable'.
+data LetNaked tyname name uni fun a = LetNaked {
+    _letAnn  :: a,
+    __letRec :: Recursivity,
+    _letBs   :: NE.NonEmpty (Binding tyname name uni fun a)
+    }
+makeLenses ''LetNaked
+
+-- a store of lets to be floated at their new position
+type FloatTable tyname name uni fun a = M.Map Pos (NE.NonEmpty (LetNaked tyname name uni fun a))
+
+-- | The 1st pass of marking floatable lets
+mark :: forall tyname name uni fun a.
+      (Ord tyname, Ord name, PLC.HasUnique tyname PLC.TypeUnique, PLC.HasUnique name PLC.TermUnique, PLC.ToBuiltinMeaning uni fun)
+     => Term tyname name uni fun a
+     -> Marks
+mark = flip runReader initCtx . go
+  where
+    initCtx :: MarkCtx
+    initCtx = (topDepth, mempty)
+
+    go :: Term tyname name uni fun a -> Reader MarkCtx Marks
+    go = breakNonRec >>> \case
+        -- lam/Lam are treated the same.
+        LamAbs _ n _ tBody  -> withLam n $ go tBody
+        TyAbs _ n _ tBody   -> withLam n $ go tBody
+
+        -- main operation: for letrec or single letnonrec
+        Let ann r bs@(representativeBindingUnique -> letU) tIn -> do
+          (depth, scope) <- ask
+          let letN = LetNaked ann r bs
+          if floatable letN
+            then do
+              let freeVars =
+                      -- if Rec, remove the here-bindings from free
+                      ifRec r (S.\\ setOf (traversed.bindingIds) bs) $
+                         calcFreeVars letN
+
+                  -- The "heart" of the algorithm: the future position to float this let to
+                  -- is determined as the maximum among its dependencies (free vars).
+                  floatPos@(floatDepth,_,_) =  maxPos $ scope `M.restrictKeys ` freeVars
+
+              -- visit the rhs'es
+              marksRhs <-
+                  -- IMPORTANT: inside the rhs, act like the current depth
+                  -- is the future floated depth of this rhs.
+                  withDepth (const floatDepth) $
+                      -- if rec, then its bindings are in scope in the rhs'es
+                      ifRec r (withBs bs floatPos) $
+                          mconcat <$> traverse go (bs^..traversed.bindingSubterms)
+
+              -- visit the inTerm
+              marksIn <-
+                  -- bindings are inscope in the InTerm for both rec&nonrec
+                  withBs bs floatPos $ go tIn
+
+              -- collect here the new mark and propagate all
+              pure $ M.singleton letU floatPos <> marksRhs <> marksIn
+            else do
+              -- since it is unfloatable (effectful), this let is a new anchor
+              let newDepth = depth+1
```

Comment:
why not: ask for the depth inside the `withDepth`? Similarly, you could ask for the scope only in the branch where you need it.